File: p109m/sorted_list_to_bst.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
const List = require('list');
const Tree = require('binary_tree');

/**
 * @param {ListNode} head
 * @return {TreeNode}
 */
var sortedListToBST = function (head) {
  if (head == null) return head;
  if (head.next == null) return new Tree.TreeNode(head.val);

  let [prev, slow, fast] = [null, head, head];
  while (fast && fast.next) {
    prev = slow;
    slow = slow.next;
    fast = fast.next.next;
  }
  prev.next = null;

  let root = new Tree.TreeNode(slow.val);
  root.left = sortedListToBST(head);
  root.right = sortedListToBST(slow.next);
  return root;
};
// TESTS
const f = (x, y) => [x, y];
[
  f([-10, -3, 0, 5, 9], '0,-3,-10,#,#,#,9,5,#,#,#'),
  f([], '#'),
  f([0], '0,#,#'),
  f([1, 3], '3,1,#,#,#'),
  f([1, 2, 3, 4], '3,2,1,#,#,#,4,#,#'),
].forEach(([array, expected]) => {
  let actual = Tree.serialize(sortedListToBST(List.fromArray(array)));
  console.log('Convert sorted list', array, 'to BST ->', actual);
  console.assert(actual === expected);
});

File: p109m/sorted_list_to_bst.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
from local_packages.list import ListNode
from local_packages.binary_tree import TreeNode


class Solution:
    def sortedListToBST(self, head: ListNode) -> TreeNode:
        if not head:
            return None
        if not head.next:
            return TreeNode(head.val)

        pslow, slow, fast = None, head, head
        while fast and fast.next:
            pslow, slow = slow, slow.next
            fast = fast.next.next
        pslow.next = None

        return TreeNode(
            slow.val,
            left=self.sortedListToBST(head),
            right=self.sortedListToBST(slow.next),
        )


# TESTS
for array, expected in [
    ([-10, -3, 0, 5, 9], "0,-3,-10,#,#,#,9,5,#,#,#"),
    ([], "#"),
    ([0], "0,#,#"),
    ([1, 3], "3,1,#,#,#"),
    ([1, 2, 3, 4], "3,2,1,#,#,#,4,#,#"),
]:
    sol = Solution()
    actual = TreeNode.serialize(sol.sortedListToBST(ListNode.from_array(array)))
    print("Convert sorted list", array, "to BST ->", actual)
    assert actual == expected
